Replace controller context

Gudahtt opened the pull request:
The `context` object previously constructed by the `ComposableController` is no more. Instead each controller now accepts its dependencies directly as constructor parameters, in a similar manner to the extension controllers.

This was done in preparation for migrating to BaseControllerV2 and the new controller messaging system - this is just a temporary solution that will let us migrate controllers one at a time.

The style of dependency injection here matches the extension (at least with newer controllers anyway). Specific methods and state snapshots are injected rather than entire controllers, to help simplify unit tests and make it easier to understand how controllers interact.

The `Engine.context` property was used throughout mobile, so it has been preserved. It is now constructed explicitly, rather than being a re-export of the `ComposableController` context.

This PR depends upon https://github.com/MetaMask/controllers/pull/387

Gudahtt committed:
Replace controller context

The `context` object previously constructed by the
`ComposableController` is no more. Instead each controller now accepts
its dependencies directly as constructor parameters, in a similar
manner to the extension controllers.

This was done in preparation for migrating to BaseControllerV2 and the
new controller messaging system - this is just a temporary solution
that will let us migrate controllers one at a time.

The style of dependency injection here matches the extension (at least
with newer controllers anyway). Specific methods and state snapshots
are injected rather than entire controllers, to help simplify unit
tests and make it easier to understand how controllers interact.

The `Engine.context` property was used throughout mobile, so it has
been preserved. It is now constructed explicitly, rather than being a
re-export of the `ComposableController` context.

This PR depends upon https://github.com/MetaMask/controllers/pull/387

Gudahtt commented:
@ibrahimtaveras00: I don't anticipate this including any functional changes, so this just needs regression testing. Unfortunately I don't have my environment setup to build mobile anymore, so I didn't test this myself :grimacing: Just wired things up equivalently to how they were before.

estebanmino commented on app/core/Engine.js:
```diff
@@ -64,70 +64,122 @@ class Engine {
+					onAssetsStateChange: listener => assetsController.subscribe(listener),
+					onPreferencesStateChange: listener => preferencesController.subscribe(listener),
+					onNetworkStateChange: listener => networkController.subscribe(listener),
+					getOpenSeaApiKey: assetsController.openSeaApiKey,
```
I was having issues with this but it worked with 

```suggestion
					getOpenSeaApiKey: () => assetsController.openSeaApiKey,
```

Gudahtt replied:
Ah, of course! :facepalm: Good catch!

Gudahtt commented on app/core/Engine.js:
```diff
@@ -64,70 +64,122 @@ class Engine {
+				),
+				new AccountTrackerController({
+					onPreferencesStateChange: listener => preferencesController.subscribe(listener),
+					initialIdentities: initialState.preferencesController?.identities
```
I found a typo - this should be `PreferencesController`, with a capital P. The identities weren't being initialized properly.

```suggestion
					initialIdentities: initialState.PreferencesController?.identities
```

Gudahtt replied:
Resolved in 31630c47601dacb53fcc110634ab6aae429d973b

Gudahtt committed:
Pass in function for `getOpenSeaApiKey` rather than string

The API key was passed in directly by accident, instead of a function get returned the key. This has been fixed.

Co-authored-by: Esteban Miño <efmino@uc.cl>

Gudahtt committed:
Update `AccountTrackerController` options

The `AccountTrackerController` option `initialIdentities` was replaced
with `getIdentities`. The initial identities passed in here were
incorrect anyway due to a typo (`initialState.preferencesController`
was used instead of `initialState.PreferencesController`).

Gudahtt commented on app/components/Views/ApproveView/Approve/index.test.js:
```diff
@@ -2,7 +2,7 @@ import React from 'react';
 import Approve from './';
 import configureMockStore from 'redux-mock-store';
-import { BN } from '@metamask/controllers';
+import { BN } from 'ethereumjs-util';
```
This change was required because `BN` is no longer exported from `@metamask/controllers` as of v9.2.0. It was removed here: https://github.com/MetaMask/controllers/pull/398/files#diff-134f671b9c1693ce708eaea4499d90b399fe55bd97be8c5b38421dcc079198faL9

So this change is unrelated to the controller context change.

Gudahtt replied:
This change has been reverted in v6.2.1, so I'm going to update this now so that this change can be removed.

Gudahtt replied:
Resolved in c45d1021318b32b0528b0c21f2529a545a9f8b2c

Gudahtt replied:
Not totally resolved - investigating now.

Gudahtt replied:
Figured it out - my attempt to update the controllers tarball was only partially successful. The lockfile still referenced the hash of the old version. This should be resolved now.

Gudahtt commented:
I've just rebased this onto `develop`, and updated the controllers bundle to match the latest version of https://github.com/MetaMask/controllers/pull/387
This is now ready for review again.

ibrahimtaveras00 commented:
I noticed as soon as I build the app I see this warning:

![Screen Shot 2021-04-12 at 2 07 22 PM](https://user-images.githubusercontent.com/41752922/114442396-82672a00-9b9a-11eb-84e3-00d314705eba.png)

ibrahimtaveras00 commented:
Tapping on account switcher shows 0 balance (I do have a balance for these 3 accounts)

![Screen Shot 2021-04-12 at 3 12 55 PM](https://user-images.githubusercontent.com/41752922/114448545-b265fb80-9ba1-11eb-9311-0ce4cf3247b8.png)

![image](https://user-images.githubusercontent.com/41752922/114448622-c9a4e900-9ba1-11eb-8bce-02121818692d.png)

Gudahtt commented:
Thanks, I'll take a look!

ibrahimtaveras00 commented:
**Issue 3:**

If I add contacts, then kill app and relaunch, those contacts get lost and I would have to re-add them

seen here = http://recordit.co/zhNMmqw5o8

Steps I took:
- go to send flow, and add a contact to address book
- I then went to settings > contacts to make sure it was there
- Kill app
- Relaunch
- double check if contacts are in the send flow and settings > contacts

Gudahtt commented on app/core/Engine.js:
```diff
@@ -64,70 +64,122 @@ class Engine {
+				),
+				new AccountTrackerController({
+					onPreferencesStateChange: listener => preferencesController.subscribe(listener),
+					getIdentities: () => preferencesController.identities
```
This line is responsible for the first error. This should be :

```suggestion
					getIdentities: () => preferencesController.state.identities
```

Gudahtt committed:
Fix `getIdentities` handler for `AccountTrackerController`

ibrahimtaveras00 commented:
**Issue 4:**

Similar to issue 3, if I add a custom token, then kill app & relaunch, the token is gone

seen here = https://recordit.co/4wFT7wQDLu

Steps I took:
- on wallet, tap add tokens
- the token I added was BLT address = 0x107c4504cd79c5d2696ea0030a8dd4e92601b82e
- kill app and relaunch

Gudahtt commented:
I'm pretty sure issues 1 & 2 are resolved with my latest update, but I don't have any guesses yet for 3 & 4. I'm going to go through this again line-by-line.

ibrahimtaveras00 commented:
Confirmed issues 1 & 2 have been resolved 👍🏽

ibrahimtaveras00 commented:
**Issue 5:**

As of latest release, when on a custom network, we don't show the basic gas modal, only advanced, but I'm seeing the basic gas estimation on this branch

seen here = https://recordit.co/2NBukLDkds

![image](https://user-images.githubusercontent.com/41752922/114466217-2b6f4e00-9bb6-11eb-8ff6-691e58723f8f.png)

Steps I took:
- I was on xDAI network, but you can also go to any default testnet 
- go through send flow, and once on confirm screen edit gas

ibrahimtaveras00 commented:
Noticed an Error:

I saw this error when attempting an Approve from https://metamask.github.io/test-dapp/

![Screen Shot 2021-04-12 at 6 16 20 PM](https://user-images.githubusercontent.com/41752922/114469672-9bcc9e00-9bbb-11eb-9a8f-53c797e8624a.png)

nonetheless the txn was confirmed 

![Screen Shot 2021-04-12 at 6 18 05 PM](https://user-images.githubusercontent.com/41752922/114469683-9f602500-9bbb-11eb-834a-32ca81f82b39.png)

ibrahimtaveras00 reviewed (changes requested):
**Issue 6:**

Currency conversion is incorrect when switching to other than USD

For instance, in this scenario I switch to AUD

![image](https://user-images.githubusercontent.com/41752922/114477483-3fbd4600-9bca-11eb-8995-edc7d03ce578.png)

![image](https://user-images.githubusercontent.com/41752922/114477406-17354c00-9bca-11eb-88c2-316813ea38a7.png)

Gudahtt committed:
Set initial controller state

The `controllers` setter on `ComposedController` used to be responsible
for setting initial state. Since that setter has been removed, the
initial state is now set after the controllers have been constructed.

This should be functionally equivalent to what it was before. We're
setting the initial state by calling `update` on each controller, just
as the `controller` setter used to.

Gudahtt commented:
I discovered that the initial controller state wasn't being set correctly in this PR. This has been fixed in 2b8023568161ad9d31b0e525f6c3318e71a436e5. This PR has also been rebased onto `develop` to resolve a small conflict.

Issues 3 & 4 should now be resolved. I'm still investigating 5 & 6.

Gudahtt commented:
Issues 5 & 6 may have been caused by the initial state bug (it looks like it was caused by a missing/invalid `chainId`). Not entirely sure though.

[The error you noticed](https://github.com/MetaMask/metamask-mobile/pull/2416#issuecomment-818280043) looks like a pre-existing issue to me (the initial custom spend limit is set to the minimum spend limit, which is optional).

Gudahtt commented on app/core/Engine.js:
```diff
@@ -64,70 +64,132 @@ class Engine {
+			// how the `ComponsedController` used to set initial state.
+			for (const controller of controllers) {
+				if (initialState[controller.name]) {
+					controller.update(this.cachedState[controller.name]);
```
Oops! `cachedState` doesn't exist. That's what the `initialState` was renamed to inside the old `ComposedController` implementation.

```suggestion
					controller.update(initialState[controller.name]);
```

Gudahtt committed:
Fix initial state variable reference

Gudahtt commented:
It turns out that the last update I made resulted in the Engine initialization crashing completely. Also it still failed to initialize controller state. It should be fixed now hopefully :crossed_fingers:

ibrahimtaveras00 reviewed (approved):
All issues have been resolved, QA Passed 👍🏽

estebanmino commented:
don't merge this yet 🙏 , we have a node bump to v14 in the making and currently working with node 10 so just to be safe we'll be merging this when that work finishes

Gudahtt commented:
Sounds good! Linking the Node v14 issue: #2205

rickycodes commented:
I just merged the node14 upgrade, so I think this should be good now.

mobularay commented:
@rickycodes can this pull request be closed? Which release is this part of, v2.2 or the upcoming v2.3?

Gudahtt commented:
This has been rebased onto `develop`. A few conflicts needed to be resolved manually in `Engine.js`. I've just reviewed the diff for `Engine.js` again to ensure all of the changes were accounted for.

mobularay commented:
yay!! Anything that needs to be done in order to close off this card? Does this need QA review, or is it good to go (to be part of v2.3 release in next sprint)?

wachunei commented:
heads up, I ran into these while trying to login:

<img width="1058" alt="Screen Shot 2021-04-26 at 19 58 58" src="https://user-images.githubusercontent.com/1024246/116165565-26d48a80-a6ca-11eb-8b0b-661e1ad6eed3.png">


<img width="1054" alt="Screen Shot 2021-04-26 at 19 59 08" src="https://user-images.githubusercontent.com/1024246/116165553-22a86d00-a6ca-11eb-9dee-710d08085860.png">

Gudahtt commented:
Interesting. At first I was going to ask that you check to ensure you _had_ a vault to unlock in that case, but... in trying to trace this, I can't figure out how the vault is supposed to get set in the KeyringController state in the first place. It looks like the KeyringController expects `vault` to be set in its initial state, but it never actually sets `vault` on its state. How did this ever work :confused:

Gudahtt commented:
Whoops, I think I found the problem. I had accidentally left the most recent commit out of that most recent rebase (07989f157e6b8224667d49d0d2d678f3ccf4008f).

I remain confused about how the `vault` state is restored, but that would explain the error you were seeing! Hopefully it works now.

estebanmino committed:
Merge branch 'develop' into replace-controller-context
